python/ray/autoscaler/node_launcher.py

```diff
@@ -13,38 +13,52 @@
 class NodeLauncher(threading.Thread):
     """Launches nodes asynchronously in the background."""
 
-    def __init__(self, provider, queue, pending, index=None, *args, **kwargs):
+    def __init__(self,
+                 provider,
+                 queue,
+                 pending,
+                 index=None,
+                 node_resource_mapping=None,
+                 *args,
+                 **kwargs):
         self.queue = queue
         self.pending = pending
         self.provider = provider
         self.index = str(index) if index is not None else ""
+        self.node_resource_mapping = node_resource_mapping
         super(NodeLauncher, self).__init__(*args, **kwargs)
 
-    def _launch_node(self, config, count, instance_type):
+    def _launch_node(self, autoscaler_config, count, node_config):
+        instance_type = self.provider.get_instance_type(node_config)
         worker_filter = {TAG_RAY_NODE_TYPE: NODE_TYPE_WORKER}
         before = self.provider.non_terminated_nodes(tag_filters=worker_filter)
-        launch_hash = hash_launch_conf(config["worker_nodes"], config["auth"])
+        launch_hash = hash_launch_conf(autoscaler_config["worker_nodes"],
+                                       autoscaler_config["auth"])
         self.log("Launching {} nodes, type {}.".format(count, instance_type))
-        node_config = config["worker_nodes"]
         node_tags = {
-            TAG_RAY_NODE_NAME: "ray-{}-worker".format(config["cluster_name"]),
+            TAG_RAY_NODE_NAME: "ray-{}-worker".format(
+                autoscaler_config["cluster_name"]),
             TAG_RAY_NODE_TYPE: NODE_TYPE_WORKER,
             TAG_RAY_NODE_STATUS: STATUS_UNINITIALIZED,
             TAG_RAY_LAUNCH_CONFIG: launch_hash,
         }
-        if instance_type:
-            node_tags[TAG_RAY_INSTANCE_TYPE] = instance_type
-            self.provider.create_node_of_type(node_config, node_tags,
-                                              instance_type, count)
-        else:
-            self.provider.create_node(node_config, node_tags, count)
+
+        node_tags[TAG_RAY_INSTANCE_TYPE] = instance_type
+        self.provider.create_node(node_config, node_tags, count)
```

Comment:
Isn't this breaking instance type creation since you need to call `create_node_of_type`?

Reply:
I think we can avoid having a create_node_of_type at all. InstanceType is always specified in the config for both the single instance and multi instance autoscalers anyways.

Reply:
But "InstanceType" is per node provider right? What you hard-coded in resource_demand_scheduler only works for AWS, whereas this special call works for any node provider that implements create_node_of_type. In other words, the configuration is per-cloud and cannot be set by the autoscaler, only the node provider.